### PR TITLE
Add portable Harbor and Verifiers exporter plugins

### LINKED ISSUE #841: Add Harbor and verifiers exporters for experiments
Add exporters so experiments can be exported in the Harbor and verifiers formats, making Kitaru experiment data usable as RL training environments/datasets.

### PR DESCRIPTION
**Pull request by strickvl**
## Summary

Kitaru users can export a frozen experiment cohort as a runnable Harbor 0.20 project or Verifiers 0.3 benchmark through separately installed exporter packages. Core Kitaru keeps the local, synchronous export operation; each exporter package supplies target-specific preflight, rendering, validation, and runtime assets.

The same typed operation remains available through `kitaru experiment export` and the standard-mode `kitaru_experiment_export` MCP tool. Both flows preserve local `source_root` and destination behavior, dry runs, bounded receipts, cancellation, safe publication, and the existing Harbor and Verifiers artifact contracts.

This deliberately does not add a worker task, REST endpoint, source or artifact transport, UI action, daemon, or Compose worker. A future worker can invoke the versioned exporter contract, but there is no backend execution path in this PR.

A companion [kitaru-template PR #10](https://github.com/zenml-io/kitaru-template/pull/10) lets the canonical returns agent run inside the generated external-evaluation environment. The returns tutorial offers export as an optional final step. Teaching `kitaru-skills` to suggest this capability remains an optional follow-up.

## Installation and user flow

Users install only the exporter they need in the same Python environment that runs the Kitaru CLI or MCP server:

```bash
uv add kitaru-harbor-exporter
# or
uv add kitaru-verifiers-exporter
```

Core discovers the selected package through the versioned `kitaru.exporters.v1` entry-point contract before remote resolution or source access. Missing, duplicate, incompatible, or unloadable providers fail with the exact package and same-environment recovery steps. Kitaru never installs plugins implicitly.

The exporter packages require Kitaru 0.23.0 or newer because 0.23.0 is the first core release with exporter contract v1. Release inventory validation and the plugin release workflow prevent either exporter from publishing against an older core.

## Architecture

| Component | Responsibility |
| --- | --- |
| Kitaru core | Remote resolution and authorization, content and environment policy, secret sanitization, immutable source capture, dependency classification, cancellation, publication, receipts, manifest contracts, and installed-exporter discovery. |
| `kitaru-harbor-exporter` | Harbor 0.20.0 preflight, one task per cohort session, agent image and adapter generation, trace ingestion, evaluator bridge, structural validation, and Harbor-native rewards and metrics. |
| `kitaru-verifiers-exporter` | Verifiers 0.3.0 preflight, one collision-safe Taskset with N Tasks, private scoring data, replaceable bundled Harness, `eval.toml`, PrimeRL 0.8.0 source configuration, and structural validation. |

The root Kitaru wheel contains no Harbor, Verifiers, or PrimeRL dependency. The exporter wheels also avoid target-framework runtime dependencies; generated artifacts declare the exact packages they need.

The Verifiers export represents a cohort as one benchmark rather than creating one environment per trace. Its Taskset and default Harness load through one artifact-specific plugin ID. A compatible external Harness can replace the default by changing `env.agent.harness.id`.

## Data and execution boundaries

Users choose what evidence leaves Kitaru. Repeatable `--omit-content`, `--include-source`, and `--exclude-source` options control exported evidence and source; `--environment-mode runtime_only` keeps ordinary registered environment values out of the artifact. Required task inputs and structural trace fields remain because the targets need them to construct and score tasks.

Attached secrets are resolved only long enough to remove their exact values from exported evidence and reject them in executable source or dependency declarations. Secret values are never serialized. Their names become runtime requirements.

Core validates every plugin manifest against the resolved evaluator IDs, policies, source exclusions, required environment names, exporter identity, target identity, and a passed structural-validation receipt before publication. Installed distribution identity is checked before Python imports the exporter entry point.

Source files are captured once into an immutable retained snapshot. Publication uses private sibling staging, cancellation checkpoints, ownership markers, destination reservations, no-replace commits, and owned rollback. Generated commands use argv rather than shell interpolation, and project-aware `uv run` commands are pinned to the dependency-classified source root.

## Reviewer Notes

Suggested review order:

1. Installed discovery and the versioned public contract in `src/kitaru/exports/plugin.py` and `render_support.py`.
2. Shared request, receipt, resolution, source-capture, and publication contracts under `src/kitaru/exports/`.
3. Harbor package behavior under `plugins/packages/harbor-exporter/` and its characterization tests.
4. Verifiers package behavior under `plugins/packages/verifiers-exporter/` and its characterization tests.
5. CLI/MCP parity, missing-package recovery, release inventory checks, and exact-target smoke.

The main review risks are the pre-import package trust check, the division between core-owned provenance and plugin-owned target output, release ordering around Kitaru 0.23.0, source time-of-check/time-of-use behavior, cancellation during publication, and separation of public Verifiers task data from private scoring evidence.

### Reproduction

Install one exporter, then dry-run an existing experiment export:

```bash
uv add kitaru-harbor-exporter

kitaru experiment export checkout-eval \
  --cohort-version checkout-errors@3 \
  --agent checkout-agent@7 \
  --format harbor \
  --source-root "$PWD" \
  --destination ./exports/checkout-harbor \
  --primary-reward correctness:exact-match:score \
  --trace-format atif \
  --trace-path /workspace/trajectory.json \
  --dry-run
```

Review the receipt, repeat without `--dry-run`, then follow the generated README. For MCP, add the exporter to the project environment that runs `uv run kitaru-mcp` and restart that subprocess.

For Verifiers, install `kitaru-verifiers-exporter`, select `--format verifiers-v1`, and run the generated project with `uv run eval @ eval.toml --model MODEL`.

See `docs/book/guides/exporting-experiments.md` for policy, dependency, provenance, Harbor, Harness replacement, and PrimeRL details.

## Validation

- `just check`
- `just test`: 3,871 passed, 16 skipped
- Plugin workspace suite: 471 passed
- Python 3.12 Harbor and Verifiers exporter tests: 53 passed
- `just plugin-artifact-smoke`
- `just mcp-schema-check`
- `just export-artifact-smoke --model-only --template-source /Users/strickvl/coding/zenml/repos/kitaru-template`
- `just export-artifact-smoke --template-source /Users/strickvl/coding/zenml/repos/kitaru-template`

The full smoke builds and installs the candidate Kitaru 0.23.0 wheel and both exporter wheels in isolated Python 3.12 environments. It proves core-only and one-plugin installation topologies, checks that core metadata contains no target dependencies, loads two generated Verifiers artifacts side by side, executes Harbor 0.20.0 in Docker, evaluates a Verifiers 0.3.0 rollout, completes the official PrimeRL 0.8.0 Env/EnvServer pre-trainer handshake, and runs the pinned canonical `python -m returns_agent.agent` entrypoint provider-free.

PrimeRL trainer, optimizer, GPU, and live model-provider execution remain outside this smoke. The generated training-source configuration and environment/orchestrator boundary are verified.

## Post-Deploy Monitoring & Validation

This is local packaging and artifact generation with no database migration or server rollout. Publish Kitaru 0.23.0 before either exporter package, then verify each exporter from a clean environment using only published wheels.

Keep the Python 3.12 exact-target job green and confirm published core metadata contains no Harbor, Verifiers, or PrimeRL dependency. Watch early reports for missing-package recovery, dependency classification, runtime requirement ownership, source exclusions, and upstream target-contract drift. Update target pins, generated configuration, documentation, and exact-target smoke together when adopting newer target versions.

---

<a href="https://app.devin.ai/review/zenml-io/kitaru/pull/703" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by htahir1** [comment]
@strickvl i like this . is this in the server?

File: src/kitaru/client/api_client.py [hunk 1]
```diff
@@ -257,17 +276,44 @@ async def request(
...
+            "auth": self._auth if authenticate else None,
+        }
+        if max_response_bytes is None:
+            response = await self._http.request(method, path, **request_kwargs)
+        else:
+            if max_response_bytes <= 0:
+                raise ValueError("max_response_bytes must be positive")
+            async with self._http.stream(method, path, **request_kwargs) as streamed:
+                declared_length = _get_content_length(streamed)
+                if declared_length is not None and declared_length > max_response_bytes:
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/client/api_client.py]: > Only use the declared length when no content encoding is present, and otherwise enforce the limit while iterating.

Fixed in `5f1d83c9`. The declared `Content-Length` precheck now runs only for unencoded responses, while `aiter_bytes()` continues to enforce the decoded-byte limit. I also added a gzip regression where the encoded length exceeds the limit but the decoded body fits exactly.

File: src/kitaru/exports/resolve.py [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+
+    working_dir = remote.agent_version.run_spec.working_dir
+    if working_dir:
+        remote._sanitizer.reject_text(
+            working_dir,
+            code="protected_value_in_path",
+            message="Protected runtime material appears in the agent working path.",
+        )
+        normalized = posixpath.normpath(working_dir.replace("\\", "/"))
+        relative = PurePosixPath(normalized)
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/exports/resolve.py:734]: > Store the normalized path in the resolved run specification before rendering.

Fixed in `5f1d83c9`. The validated POSIX-normalized path is now copied into the resolved agent run specification before either renderer receives it. Regression coverage includes both `src\\agent` and `missing/../src`.

**Comment by strickvl** [comment]
> is this in the server?

Not as a new REST export endpoint. The export operation lives in core Kitaru and uses the existing server APIs to resolve the experiment, cohort, agent, sessions, evaluators, and secrets. It then snapshots the user's local source and generates the Harbor or Verifiers project locally. Users can invoke that same operation through the CLI or through the standard-mode MCP server tool.

**Comment by htahir1** [comment]
@strickvl id model it as a REST endpoint so you can download a zip file from the UI if possible?

File: plugins/packages/verifiers-exporter/src/kitaru_verifiers_exporter/exporter.py [hunk 3]
```diff
@@ -0,0 +1,1180 @@
...
+            "OPENAI_MODEL": ctx.model,
+        }
+        existing_path = env.get("PATH", "/usr/local/bin:/usr/bin:/bin")
+        env["PATH"] = (_RUNTIME_ROOT / ".venv" / "bin").as_posix() + ":" + existing_path
+        return await runtime.run_program(
+            [
+                "/usr/bin/env",
+                "-C",
+                (_RUNTIME_ROOT / _RUNTIME["working_directory"]).as_posix(),
+                *_RUNTIME["command_argv"],
```
[the remark this answers is not in the document]
Comment (strickvl) [reply plugins/packages/verifiers-exporter/src/kitaru_verifiers_exporter/exporter.py:331]: > Pin uv run to the classified root project

Addressed: generated Verifiers runtimes now pin project-aware `uv run` commands to `/workspace/kitaru-agent`. Harbor applies the same rule against its copied project root, `/workspace`. Both preserve an explicit `--no-project`, and regression coverage includes nested working directories and conflicting explicit `--project` values.

File: src/kitaru/exports/formats/harbor.py [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+        if result.return_code != 0:
+            raise RuntimeError(
+                f"Exported agent failed with exit code {result.return_code}"
+            )
+        destination = "/logs/agent/" + runtime["trace_log_name"]
+        copied = await environment.exec(
+            "mkdir -p /logs/agent && cp -- "
+            + shlex.quote(runtime["trace_path"])
+            + " "
+            + shlex.quote(destination),
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/exports/formats/harbor.py]: > Skip copying a Harbor trace onto itself

Addressed: the generated Harbor agent now verifies a trace already at the canonical log path in place and only copies when source and destination differ. Regression coverage exercises both ATIF and Kitaru canonical paths.

**Comment by strickvl** [comment]
> id model it as a REST endpoint so you can download a zip file from the UI if possible?

We thought this through and kept this PR local-first. The core export operation now discovers an installed Harbor or Verifiers exporter package, but CLI and MCP still execute it synchronously against the user's local source and destination. A server job would also require source upload, artifact storage/download, worker routing, and a new authorization/lifecycle contract, so we're deferring that until there is a concrete UI or remote-execution use case. The plugin contract is caller-neutral, so a future worker job can invoke the same exporter without another package redesign.

File: src/kitaru/client/api_client.py [hunk 5]
```diff
@@ -257,17 +277,48 @@ async def request(
...
+                    if "Content-Encoding" not in streamed.headers
+                    else None
+                )
+                if declared_length is not None and declared_length > max_response_bytes:
+                    raise ResponseTooLargeError(
+                        max_response_bytes, content_length=declared_length
+                    )
+                body = bytearray()
+                async for chunk in streamed.aiter_bytes():
+                    if len(body) + len(chunk) > max_response_bytes:
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/client/api_client.py]: > Stream raw bytes through an incremental decoder with a bounded output size.

Addressed in `e0ea72cd`: bounded responses now read raw HTTPX chunks and incrementally decode gzip, x-gzip, and both deflate forms with a strict decoded-output cap. Unsupported or stacked encodings fail before the body is read, and truncated compressed streams fail instead of returning partial data. The full repository suite passes with 3,856 tests and 16 skips.

File: src/kitaru/exports/_runtime.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        "fish",
+        "ksh",
+        "powershell",
+        "pwsh",
+        "sh",
+        "tcsh",
+        "zsh",
+    }
+)
+_UV_FLAGS_WITH_VALUE = frozenset({"--directory", "--package", "--project", "--python"})
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/exports/_runtime.py]: > Reject uv commands that override the Python runtime

Addressed in `d3e8e907`: core export preflight now rejects uv interpreter selection before either exporter runs, including `--python VALUE`, `--python=VALUE`, `-p VALUE`, and joined short forms. Regression coverage confirms the sibling Python runtime controls remain unsupported.

File: src/kitaru/exports/config.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    format: ExportFormat
+    source_root: Path
+    destination: Path
+    primary_reward: str
+    required_environment_names: tuple[str, ...] = Field(default=(), max_length=100)
+    content_policy: ContentPolicy = Field(default_factory=ContentPolicy)
+    environment_policy: EnvironmentPolicy = Field(default_factory=EnvironmentPolicy)
+    source_policy: SourcePolicy = Field(default_factory=SourcePolicy)
+    trace_format: TraceFormat | None = None
+    trace_path: str | None = None
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/exports/config.py]: > Validate trace-path length at the request boundary

Addressed in `d3e8e907`: the 1,024-character boundary is now enforced consistently by the shared export request and MCP request model. Oversized paths produce `invalid_arguments` through both CLI and MCP, with boundary and schema regression coverage.

File: src/kitaru/client/api_client.py [hunk 8]
```diff
@@ -70,6 +72,119 @@
...
+            try:
+                chunk = self._decompressor.decompress(pending, available + 1)
+            except zlib.error as error:
+                raise httpx.DecodingError("Invalid compressed response body") from error
+            if len(chunk) > available:
+                raise ResponseTooLargeError(limit)
+            decoded.extend(chunk)
+            unconsumed = self._decompressor.unconsumed_tail
+            if not unconsumed:
+                break
```
[the remark this answers is not in the document]
Comment (strickvl) [reply src/kitaru/client/api_client.py:189]: > Decode every member in gzip responses

Addressed in `d3e8e907`: after each gzip member reaches EOF, the decoder passes `unused_data` to a fresh gzip decompressor and continues under the same cumulative decoded-byte limit. Regressions cover valid within-limit concatenated responses and over-limit expansion across members.
